server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java

```diff
@@ -810,6 +796,7 @@ public DataStream snapshot(Collection<String> indicesInSnapshot) {
             lifecycle,
             failureStore,
             failureIndices,
+            rolloverOnWrite,
```

Comment:
I wasn't 100% sure about the snapshot and restore ones (see the change in the `RestoreService.java`). I don't immediately see a reason why we wouldn't want to copy the flag when snapshotting or restoring, so I decided to keep it in (rather than overriding it to `false`). Let me know if I'm missing something here.

Reply:
I am leaning also into copying it. If this flag was added because something changes in the index template, I think it's safer to apply the rollover at the next write to ensure we have the latest index template data.

Reply:
If we restore a data stream as a different name that no longer matches the template we wont be able to roll it over. We document that this as a concern currently: https://www.elastic.co/guide/en/elasticsearch/reference/current/restore-snapshot-api.html#restore-snapshot-api-prereqs

Restoring the data stream without a template poses a maintenance problem, but if we restore a data stream in that way with the rollover on write flag set, will it cause us to not be able to write to the data stream at all until it is remedied? That would take that corner case from being just a maintenance problem to a possible availability problem.

I don't think this is a blocker for merging this, but I think we may want to make sure that kind of interaction is what we expect it to be and note it as a "gotcha" in the documentation.

Reply:
Ah that's an excellent point. I don't immediately see an easy/sensible way to avoid that "gotcha" while keeping the `rolloverOnWrite` flag set to true (I see some ways, but they're either very limiting in the snapshot/restore process or they require elaborate changes in the rollover process, both of which I'm not a big fan of right away). I think I'm still on board with copying the flag, but I agree that we should make this "gotcha" (more) clear in the documentation.

Reply:
I wanted to include the doc change in this PR, see [59b66a7](https://github.com/elastic/elasticsearch/pull/107122/commits/59b66a73155a9122b11f50ef6810135cc3010943). Let me know if either of you had anything else in mind. (fingers crossed that this renders like I intended)

Reply:
Hm, very good point:

> That would take that corner case from being just a maintenance problem to a possible availability problem.

I think this is not very good user experience.

What if we checked if there is a valid template at the moment of the restoration, we have the cluster state available at: `https://github.com/elastic/elasticsearch/blob/882b92ab6092a8ea6b53a64e90cb12e812241483/server/src/main/java/org/elasticsearch/snapshots/RestoreService.java#L1409` 

If there is a template we copy the `rolloverOnWrite`, if not we set it to `false`.  We could even add an info log line restoring a data stream without a template, so we can debug this.

If you think this is too complex, then I am in favour of not copying the flag, and give the user the recommendation to rollover if they want to use the current index template.

What are your thoughts?

Reply:
> If there is a template we copy the `rolloverOnWrite`, if not we set it to `false`. We could even add an info log line restoring a data stream without a template, so we can debug this.

Ah I like this idea. I don't have an issue with the complexity myself, but I am trying to figure out what impact this would have on the UX as well. Ideally, I think I'd like to make this clear to the user in some way other than a log line as well (to hopefully reduce SDHs coming our way about this).

If we do not copy the `rolloverOnWrite` flag, and a user creates a matching template afterwards (with a different mapping or w/e), we won't automatically roll over because it isn't considered a "change" ([source](https://github.com/nielsbauman/elasticsearch/blob/59b66a73155a9122b11f50ef6810135cc3010943/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/IndexTemplateRegistry.java#L550-L550)). I guess this would be a less "blocking" flow than the current behaviour in this PR, but I'm trying to come up with a way to make this clear to the user. Maybe we could add a `warn` log line along the lines of
> Restoring a data stream that was marked for lazy rollover, without a matching index template. The restored data stream will NOT be marked for lazy rollover, and might thus not include the latest changes in the index template that caused the original data stream to be marked for rollover.

Reply:
I think the fact that you can restore a data stream without a template is vaguely bug-like already considering that the templates are so tightly coupled with how data streams are managed.

I think we could potentially surface a warning in the logs as well as via the `HeaderWarning` internal API, but lets discuss some more off this PR.